File: ios/Releaf/Data/Shelf/ShelfEntity.swift
/*
 * ShelfEntity.swift
 *
 * GRDB record for the `shelves` table — top of the Shelf → Book →
 * Chapter → Page hierarchy. Introduced in GRDB migration v3; existing
 * notebooks land on the default "General" shelf seeded by the same
 * migration.
 */

import Foundation
import GRDB

public struct ShelfEntity: Codable, FetchableRecord, PersistableRecord,
                           Identifiable, Equatable, Sendable {

    public static let databaseTableName = "shelves"

    public var id: String
    public var name: String
    /// Hex color (e.g. `#7AA874`) or nil for theme default.
    public var colorHex: String?
    /// Manual ordering hint. 1024-step spacing leaves room for
    /// re-ordering without renumbering neighbours.
    public var position: Int64
    public var createdAt: String
    public var updatedAt: String
    public var dirty: Bool
    public var deletedAt: String?

    public enum CodingKeys: String, CodingKey {
        case id
        case name
        case colorHex  = "color_hex"
        case position
        case createdAt = "created_at"
        case updatedAt = "updated_at"
        case dirty
        case deletedAt = "deleted_at"
    }

    public init(
        id: String,
        name: String,
        colorHex: String? = nil,
        position: Int64 = 1024,
        createdAt: String,
        updatedAt: String,
        dirty: Bool = true,
        deletedAt: String? = nil
    ) {
        self.id = id
        self.name = name
        self.colorHex = colorHex
        self.position = position
        self.createdAt = createdAt
        self.updatedAt = updatedAt
        self.dirty = dirty
        self.deletedAt = deletedAt
    }

    /// Stable id of the default "General" shelf seeded by migration.
    public static let defaultGeneralId: String = "shelf-general"
}

File: apps/releaf/android/app/src/main/java/app/releaf/mobile/data/shelf/ShelfEntity.kt
/*
 * ShelfEntity.kt
 *
 * Room entity for the `shelves` table — top of the Shelf → Book →
 * Chapter → Page hierarchy. A shelf is a user-editable container
 * (name, color, position); books live inside exactly one shelf.
 *
 * Introduced in DB v12. Existing notebooks are backfilled onto the
 * default `shelf-general` shelf by Migration11To12.
 */

package app.releaf.mobile.data.shelf

import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.Index
import androidx.room.PrimaryKey

@Entity(
    tableName = "shelves",
    indices = [
        Index("position"),
        Index("deleted_at"),
    ],
)
data class ShelfEntity(
    @PrimaryKey
    @ColumnInfo(name = "id")
    val id: String,

    @ColumnInfo(name = "name")
    val name: String,

    /** Hex color (e.g. `#7AA874`) or null for theme default. */
    @ColumnInfo(name = "color_hex")
    val colorHex: String? = null,

    /** 1024-step spacing leaves room for reordering without renumbering. */
    @ColumnInfo(name = "position", defaultValue = "1024")
    val position: Long = 1024L,

    @ColumnInfo(name = "created_at")
    val createdAt: String,

    @ColumnInfo(name = "updated_at")
    val updatedAt: String,

    @ColumnInfo(name = "dirty", defaultValue = "1")
    val dirty: Boolean = true,

    @ColumnInfo(name = "deleted_at")
    val deletedAt: String? = null,
) {
    companion object {
        /** Stable id of the default "General" shelf seeded by migration. */
        const val DEFAULT_GENERAL_ID = "shelf-general"
    }
}
